webui/react/src/components/kit/CodeEditor.tsx

```diff
@@ -0,0 +1,367 @@
+
+const { DirectoryTree } = Tree;
+
+import css from './CodeEditor/CodeEditor.module.scss';
+
+import './CodeEditor/index.scss';
+
+export type Props = {
+  files: TreeNode[];
+  onSelectFile?: (arg0: string) => void;
+  readonly?: boolean;
+  selectedFilePath?: string;
+};
+
+const DEFAULT_DOWNLOAD_INFO = {
+  fileName: '',
+  url: '',
+};
+
+const sortTree = (a: TreeNode, b: TreeNode) => {
+  if (a.children) a.children.sort(sortTree);
+
+  if (b.children) b.children.sort(sortTree);
+  // sorting first by having an extension or not, then by extension first
+  // and finally alphabetically.
+  const titleA = String(a.title);
+  const titleB = String(b.title);
+
+  if (isConfig(a.key) && !isConfig(b.key)) return -1;
+  if (!isConfig(a.key) && isConfig(b.key)) return 1;
+  // submitted before runtime for whatever reason
+  if (isConfig(a.key) && isConfig(b.key)) return a.key < b.key ? 1 : -1;
+
+  if (!a.isLeaf && b.isLeaf) return -1;
+
+  if (a.isLeaf && !b.isLeaf) return 1;
+
+  if (!a.isLeaf && !b.isLeaf) return titleA.localeCompare(titleB) - titleB.localeCompare(titleA);
+
+  // had to use RegEx due to some files being ".<filename>"
+  const [stringA, extensionA] = titleA.split(/^(?=[a-zA-Z])\./);
+  const [stringB, extensionB] = titleB.split(/^(?=[a-zA-Z])\./);
+
+  if (!extensionA && extensionB) return 1;
+
+  if (!extensionB && extensionA) return -1;
+
+  if (!extensionA && !extensionB)
+    return stringA.localeCompare(stringB) - stringB.localeCompare(stringA);
+
+  return extensionA.localeCompare(extensionB) - extensionB.localeCompare(extensionB);
+};
+
+const PageError = {
+  Decode: 'Could not decode file.',
+  Empty: 'File has no content.',
+  Fetch: 'Unable to fetch file.',
+  None: '',
+} as const;
+
+type PageError = ValueOf<typeof PageError>;
+
+const Config = {
+  Runtime: 'Runtime Configuration',
+  Submitted: 'Submitted Configuration',
+} as const;
+
+type Config = ValueOf<typeof Config>;
+
+const descForConfig = {
+  [Config.Submitted]: 'original submitted config',
+  [Config.Runtime]: 'after merge with defaults and templates',
+};
+
+const isConfig = (key: unknown): key is Config =>
+  key === Config.Submitted || key === Config.Runtime;
+
+/**
+ * A component responsible to enable the user to view the code for a experiment.
+ *
+ * It renders a file tree and a selected file in the MonacoEditor
+ *
+ * Props:
+ *
+ * files: an array of one or more files to display code;
+ *
+ * onSelectFile: called with filename when user changes files in the tree;
+ *
+ * readonly: prevent user from making changes to code files;
+ *
+ * selectedFilePath: gives path to the file to set as activeFile;
+ */
+
+const CodeEditor: React.FC<Props> = ({ files, onSelectFile, readonly, selectedFilePath }) => {
+  const [pageError, setPageError] = useState<PageError>(PageError.None);
+
+  const [activeFile, setActiveFile] = useState<TreeNode | null>(files.length > 0 ? files[0] : null);
```

Comment:
When I wrote it, I was worried about trying to access `files[0]` if files is 0-length or undefined but it looks like now that we require files param, it would be OK